File: remote/src/main/java/com/kelsos/mbrc/interactors/VolumeInteractor.java
package com.kelsos.mbrc.interactors;

import com.google.inject.Inject;
import com.kelsos.mbrc.dto.player.Volume;
import com.kelsos.mbrc.dto.requests.VolumeRequest;
import com.kelsos.mbrc.services.api.PlayerService;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class VolumeInteractor {
  @Inject private PlayerService api;

  public Observable<Integer> getVolume() {
    return api.getVolume().map(Volume::getValue).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
  }

  public Observable<Integer> setVolume(int volume) {
    return api.updateVolume(new VolumeRequest().setValue(volume))
        .map(Volume::getValue)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/VolumeInteractor.kt
package com.kelsos.mbrc.interactors

import com.kelsos.mbrc.dto.requests.VolumeRequest
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.services.api.PlayerService
import rx.Observable
import javax.inject.Inject

class VolumeInteractor
@Inject constructor(private val api: PlayerService) {

  fun getVolume(): Observable<Int> {
    return api.getVolume()
        .map { it.value }
        .task()
  }

  fun setVolume(volume: Int): Observable<Int> {
    val volumeRequest = VolumeRequest()
    volumeRequest.value = volume
    return api.updateVolume(volumeRequest)
        .map { it.value }
        .task()
  }
}
